inference-engine/src/mkldnn_plugin/nodes/mkldnn_roi_align_node.cpp

```diff
@@ -0,0 +1,348 @@
+            for (int xBinInd = 0; xBinInd < pooledW; ++xBinInd) {
+                // run into bin
+                for (int ySampleInd = 0; ySampleInd < samplingRatioY;
+                     ySampleInd++) {
+                    float sampleY = y1 + yBinInd * binHeight +
+                                     sampleDistanceY * (0.5f + ySampleInd);
+                    for (int xSampleInd = 0; xSampleInd < samplingRatioX;
+                         xSampleInd++) {
+                        float sampleX = x1 + xBinInd * binWidth +
+                                         sampleDistanceX * (0.5f + xSampleInd);
+
+                        if (sampleX < -1.0 || sampleX > W ||
+                            sampleY < -1.0 || sampleY > H) {
+                            // For this sample we save 4x point (0,0) with weight 0
+                            pointVector.insert(pointVector.end(), 4, {0, 0});
+                            weightVector.insert(weightVector.end(), 4, float{0});
+                            continue;
+                        }
+                        sampleX = std::max(sampleX, float{0});
+                        sampleY = std::max(sampleY, float{0});
+
+                        auto sampleYLow = static_cast<unsigned int>(sampleY);
+                        auto sampleXLow = static_cast<unsigned int>(sampleX);
+                        unsigned int sampleYHigh;
+                        unsigned int sampleXHigh;
+                        if (sampleYLow >= H - 1) {
+                            sampleYHigh = sampleYLow = H - 1;
+                            sampleY = static_cast<float>(sampleYLow);
+                        } else {
+                            sampleYHigh = sampleYLow + 1;
+                        }
+                        if (sampleXLow >= H - 1) {
+                            sampleXHigh = sampleXLow = W - 1;
+                            sampleX = static_cast<float>(sampleXLow);
+                        } else {
+                            sampleXHigh = sampleXLow + 1;
+                        }
+                        pointVector.push_back({sampleYLow, sampleXLow});
+                        pointVector.push_back({sampleYLow, sampleXHigh});
+                        pointVector.push_back({sampleYHigh, sampleXLow});
+                        pointVector.push_back({sampleYHigh, sampleXHigh});
+
+                        // weight calculation for bilinear interpolation
+                        auto ly = sampleY - sampleYLow;
+                        auto lx = sampleX - sampleXLow;
+                        auto hy = 1.0f - ly;
+                        auto hx = 1.0f - lx;
+
+                        weightVector.push_back(hy * hx);
+                        weightVector.push_back(hy * lx);
+                        weightVector.push_back(ly * hx);
+                        weightVector.push_back(ly * lx);
+                    }
+                }
+            }
+        }
+        parallel_for(C, [&](int c) {
```

Comment:
This for sure will be inefficient for Blocked and mostly for NHWC layouts because of bad memory access pattern.
For NHWC you need to put cycle by C after pooledH and pooledW cycles. Parallelization in this case should be done by pooledH, pooledW.
For Blocked layout it also makes sense to split C cycle into two: 1) "by C blocks" which will keep be outermost (except Rois) 2) "inside the block" which should be placed after pooledH and pooledW. Parallelization in this case may involve 3 cycles: C Blocks, pooledH, pooledW

Reply:
Done